src/dodal/devices/i04/murko_results.py

```diff
@@ -0,0 +1,221 @@
+            "most_likely_click"
+        ]  # As proportion from top, left of image
+        shape = closest_result["original_shape"]  # Dimensions of image in pixels
+        # Murko returns coords as y, x
+        centre_px = (coords[1] * shape[1], coords[0] * shape[0])
+        LOGGER.info(
+            f"Using image taken at {closest_omega}, which found xtal at {centre_px}"
+        )
+
+        beam_dist_px = calculate_beam_distance(
+            self.get_beam_centre(),  # beam centre
+            centre_px[0],
+            centre_px[1],
+        )
+
+        return camera_coordinates_to_xyz_mm(
+            beam_dist_px[0],
+            beam_dist_px[1],
+            closest_omega,
+            metadata["microns_per_x_pixel"],
+            metadata["microns_per_y_pixel"],
+        )
+
+    @AsyncStatus.wrap
+    async def trigger(self):
+        # May be better as kickoff/complete?
+        # Wait for results
+        sample_id = await self.sample_id.get_value()
+        next_angle = self.angles_to_search.pop(0)
+        final_message = None
+        while next_angle:  # Goes round in a loop until 270 degrees is found
+            # waits here for next batch to be recieved (10 scans?). pickle deserialises data
+            message = await self.pubsub.get_message(timeout=self.TIMEOUT_S)
+            if message is None:  # No more messages to process
+                await self.process_batch(
+                    final_message, sample_id, next_angle
+                )  # Process final message again
+                break
+            next_angle = await self.process_batch(message, sample_id, next_angle)
+            final_message = message
+        x_values = list(self.coords["x"].values())
+        y_values = list(self.coords["y"].values())
+        z_values = list(self.coords["z"].values())
+        assert x_values, "No x values"
+        assert z_values, "No z values"
+        assert y_values, "No y values"
+        self._x_mm_setter(np.mean(x_values))
+        self._y_mm_setter(np.mean(y_values))
+        self._z_mm_setter(np.mean(z_values))
+
+    async def process_batch(self, message, sample_id, next_angle):
+        if message and message["type"] == "message":
+            batch_results = pickle.loads(message["data"])
+            for results in batch_results:
+                print(f"Got {results} from redis")
+                assert isinstance(results, dict)
+                for uuid, result in results.items():
+                    if metadata_str := await self.redis_client.hget(
+                        f"murko:{sample_id}:metadata", uuid
+                    ):
+                        next_angle = self.process_result(
+                            result, uuid, metadata_str, next_angle
+                        )
+        return next_angle
+
+    def process_result(
+        self, result: dict, uuid: int, metadata_str: str, search_angle: float | None
+    ) -> float:
+        if search_angle is None:
+            return None
```

Comment:
Should: The return type on here should be `float | None`, I'm surprised the type system is happy how it is